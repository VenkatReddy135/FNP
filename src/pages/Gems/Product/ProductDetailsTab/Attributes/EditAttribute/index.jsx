/* eslint-disable react/jsx-props-no-spreading */
import React, { useState, useEffect } from "react";
import { TextInput, required, SimpleForm, useTranslate, useMutation, useNotify } from "react-admin";
import PropTypes from "prop-types";
import { Grid, Typography, Divider } from "@material-ui/core";
import SimpleModel from "../../../../../../components/CreateModal";
import CommonDialogContent from "../../../../../../components/CommonDialogContent";
import SwitchComp from "../../../../../../components/switch";
import CustomToolbar from "../../../../../../components/CustomToolbar";
import { useCustomQueryWithStore, useBoolean, onFailure, onSuccess } from "../../../../../../utils/CustomHooks";
import formatDateValue from "../../../../../../utils/formatDateTime";
import useStyles from "../../../../../../assets/theme/common";
import CustomViewUI from "../../../../../../components/CustomViewUI/CustomViewUI";
import { TIMEOUT } from "../../../../../../config/GlobalConfig";
import LoaderComponent from "../../../../../../components/LoaderComponent";

/**
 * Component for edit attribute UI
 *
 * @param {object} props all props required
 * @param {object} match to get the id
 * @param {object} history to get the router API's for navigations
 * @returns {React.ReactElement} returns a edit attribute UI
 */
const EditAttribute = (props) => {
  const { match, history } = props;
  const translate = useTranslate();
  const classes = useStyles();
  const requiredValidate = [required()];
  const notify = useNotify();
  const [mutate] = useMutation();
  const [modalOpenFlag, isModalOpen] = useBoolean(false);
  const [isEnabled, setEnabled] = useState(true);
  const [responseData, setResponseData] = useState({});
  const { createdBy, createdAt, updatedBy, updatedAt, isEnabled: enabledValue } = responseData;
  const [editObj, setEditObj] = useState({
    attributeName: "",
    attributeValue: "",
  });
  const productId = match?.params?.id;

  /**
   * @function handleAttributeInfo This function will set attribute details of id
   * @param {object} response is passed to the function
   */
  const handleAttributeInfo = (response) => {
    setResponseData(response?.data);
  };
  const resourceAttributeInfo = `${window.REACT_APP_GEMS_SERVICE}products/attributes/${productId}`;
  const { loading } = useCustomQueryWithStore("getOne", resourceAttributeInfo, handleAttributeInfo);

  /**
   * @function handleSubmit to submit a form onClick event
   * @param {object} formValues object to get the response status
   */
  const handleSubmit = (formValues) => {
    const { isEnabled: i, attributeName: aN, attributeValue: aV, enabled: e } = formValues;
    isModalOpen.on();
    setEditObj({ isEnabled: i, attributeName: aN, attributeValue: aV, enabled: e });
  };

  /**
   * @function handleUpdateSuccess This function will handle Success on Update
   * @param {object} response object to get the response status
   */
  const handleUpdateSuccess = (response) => {
    notify(response?.data?.message || translate("update_attribute_success"), "info", TIMEOUT);
    isModalOpen.off();
    history.goBack();
  };

  /**
   * @function updateTagRelation function called when button click for update
   */
  const updateAttribute = () => {
    mutate(
      {
        type: "put",
        resource: `${window.REACT_APP_GEMS_SERVICE}products/attributes/${productId}`,
        payload: {
          data: {
            attributeName: editObj?.attributeName,
            attributeValue: editObj?.attributeValue,
            isEnabled,
            enabled: responseData?.enabled,
          },
        },
      },
      {
        onSuccess: (response) => {
          onSuccess({ response, notify, translate, handleSuccess: handleUpdateSuccess });
        },
        onFailure: (error) => {
          onFailure({ error, notify, translate });
        },
      },
    );
  };

  useEffect(() => {
    setEnabled(enabledValue);
  }, [enabledValue]);

  /**
   * @function handleCancel to cancel the updation of attribute details
   */
  const handleCancel = () => {
    history.goBack();
  };

  /**
   * @param {boolean} value to set in the state
   * @function handleIsEnabledChange to change the enable flags
   */
  const handleIsEnabledChange = (value) => {
    setEnabled(value);
  };

  return (
    <>
      <Grid container direction="row" justify="space-between">
        <Typography variant="h5" data-testid="edit_attribute" className={classes.gridStyle}>
          {translate("edit_attribute")}
        </Typography>
      </Grid>
      <Divider variant="fullWidth" className={classes.dividerStyle} />
      {loading ? (
        <LoaderComponent />
      ) : (
        <SimpleForm
          save={handleSubmit}
          toolbar={<CustomToolbar onClickCancel={handleCancel} saveButtonLabel={translate("update")} />}
          initialValues={responseData}
        >
          <>
            <Grid container>
              <Grid item xs={3}>
                <CustomViewUI
                  label={translate("attribute_type")}
                  value={responseData?.attributeType?.name}
                  gridSize={{ xs: 12 }}
                />
              </Grid>
              <Grid item xs={3}>
                <TextInput
                  source="attributeName"
                  label={translate("attribute_name")}
                  className={classes.textInputField}
                  validate={requiredValidate}
                  data-testid="attribute_name"
                />
              </Grid>
              <Grid item xs={3}>
                <TextInput
                  source="attributeValue"
                  label={translate("attribute_value")}
                  className={classes.textInputField}
                  validate={requiredValidate}
                  data-testid="attribute_value"
                />
              </Grid>
            </Grid>
            <Grid container data-testid="is_enabled" direction="row">
              <Grid item xs={12}>
                <Typography variant="caption">{translate("is_enabled")}</Typography>
              </Grid>
              <Grid item xs={12}>
                <SwitchComp record={isEnabled} onChange={handleIsEnabledChange} disable={false} />
              </Grid>
            </Grid>
          </>
          <Grid container className={classes.listStyle}>
            <Grid item data-testid="relations_created_by" md={3}>
              <CustomViewUI label={translate("relations_created_by")} value={createdBy} />
            </Grid>
            <Grid item data-testid="relations_created_date" md={9}>
              <CustomViewUI
                label={translate("relations_created_date")}
                value={createdAt ? formatDateValue(createdAt) : null}
              />
            </Grid>
          </Grid>
          <Grid container className={classes.tabStyle}>
            <Grid item data-testid="last_modified_by" md={3}>
              <CustomViewUI label={translate("last_modified_by")} value={updatedBy} />
            </Grid>
            <Grid item data-testid="last_modified_date" md={9}>
              <CustomViewUI
                label={translate("last_modified_date")}
                value={updatedAt ? formatDateValue(updatedAt) : null}
              />
            </Grid>
          </Grid>
        </SimpleForm>
      )}
      <SimpleModel
        dialogContent={<CommonDialogContent message={translate("attribute_update_message")} />}
        showButtons
        closeText={translate("cancel")}
        actionText={translate("continue")}
        openModal={modalOpenFlag}
        handleClose={isModalOpen.off}
        handleAction={updateAttribute}
      />
    </>
  );
};

EditAttribute.propTypes = {
  match: PropTypes.objectOf(PropTypes.any),
  history: PropTypes.objectOf(PropTypes.any),
};
EditAttribute.defaultProps = {
  match: {},
  history: {},
};

export default EditAttribute;
