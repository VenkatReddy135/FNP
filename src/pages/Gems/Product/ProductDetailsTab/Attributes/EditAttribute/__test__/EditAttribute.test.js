/* eslint-disable react/jsx-props-no-spreading */
import React from "react";
import { cleanup, render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import EditAttribute from "../index";

jest.mock("react-admin", () => ({
  useNotify: jest.fn(),
  useMutation: () => [() => {}],
  useQueryWithStore: () => {
    return { loading: false };
  },
  useRefresh: () => {},
  required: () => true,
  useTranslate: () => (label) => label,
  SimpleForm: "SimpleForm",
  TextInput: "TextInput",
}));

describe("Feature List test cases", () => {
  const props = {
    match: {
      params: {
        id: "P_00000133",
      },
    },
  };
  beforeEach(() => {
    render(<EditAttribute {...props} />);
  });

  afterEach(() => {
    cleanup();
  });

  it("should render list component", () => {
    expect(screen).toBeTruthy();
  });

  it("should include title", () => {
    const label = screen.getByTestId("edit_attribute");
    expect(label).toBeTruthy();
  });

  it("should include title", () => {
    const label = screen.getByTestId("edit_attribute");
    expect(label).toBeTruthy();
  });

  it("should include attribute_name", () => {
    const label = screen.getByTestId("attribute_name");
    expect(label).toBeTruthy();
  });

  it("should include attribute_value", () => {
    const label = screen.getByTestId("attribute_value");
    expect(label).toBeTruthy();
  });

  it("should include is_enabled", () => {
    const label = screen.getByTestId("is_enabled");
    expect(label).toBeTruthy();
  });

  it("should include relations_created_by", () => {
    const label = screen.getByTestId("relations_created_by");
    expect(label).toBeTruthy();
  });

  it("should include relations_created_date", () => {
    const label = screen.getByTestId("relations_created_date");
    expect(label).toBeTruthy();
  });

  it("should include last_modified_by", () => {
    const label = screen.getByTestId("last_modified_by");
    expect(label).toBeTruthy();
  });

  it("should include last_modified_date", () => {
    const label = screen.getByTestId("last_modified_date");
    expect(label).toBeTruthy();
  });

  it("should match the snapshot", () => {
    const snap = renderer.create(<EditAttribute {...props} />);
    expect(snap.toJSON()).toMatchSnapshot();
  });
});
