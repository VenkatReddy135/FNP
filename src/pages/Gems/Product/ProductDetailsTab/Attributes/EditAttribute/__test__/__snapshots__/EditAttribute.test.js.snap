// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature List test cases should match the snapshot 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-justify-xs-space-between"
  >
    <h5
      className="MuiTypography-root makeStyles-gridStyle-1142 MuiTypography-h5"
      data-testid="edit_attribute"
    >
      edit_attribute
    </h5>
  </div>,
  <hr
    className="MuiDivider-root makeStyles-dividerStyle-1156"
  />,
  <SimpleForm
    initialValues={Object {}}
    save={[Function]}
    toolbar={
      <CustomToolbar
        disabled={false}
        onClickCancel={[Function]}
        onClickSubmit={[Function]}
        saveButtonIcon={<React.Fragment />}
        saveButtonLabel="update"
      />
    }
  >
    <div
      className="MuiGrid-root MuiGrid-container"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-12"
        >
          <span
            className="MuiTypography-root MuiTypography-caption"
          >
            attribute_type
          </span>
          <h6
            className="MuiTypography-root MuiTypography-h6"
          />
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <TextInput
          className="makeStyles-textInputField-1154"
          data-testid="attribute_name"
          label="attribute_name"
          source="attributeName"
          validate={
            Array [
              true,
            ]
          }
        />
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <TextInput
          className="makeStyles-textInputField-1154"
          data-testid="attribute_value"
          label="attribute_value"
          source="attributeValue"
          validate={
            Array [
              true,
            ]
          }
        />
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-container"
      data-testid="is_enabled"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <span
          className="MuiTypography-root MuiTypography-caption"
        >
          is_enabled
        </span>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <span
          className="MuiSwitch-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1218 MuiSwitch-switchBase PrivateSwitchBase-checked-1219 Mui-checked"
            data-at-id="switch"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                checked={true}
                className="PrivateSwitchBase-input-1221 MuiSwitch-input"
                onChange={[Function]}
                type="checkbox"
              />
              <span
                className="MuiSwitch-thumb"
              />
            </span>
          </span>
          <span
            className="MuiSwitch-track"
          />
        </span>
      </div>
    </div>
    <div
      className="MuiGrid-root makeStyles-listStyle-1159 MuiGrid-container"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-3"
        data-testid="relations_created_by"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 MuiGrid-grid-md-4"
        >
          <span
            className="MuiTypography-root MuiTypography-caption"
          >
            relations_created_by
          </span>
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            -
          </h6>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-9"
        data-testid="relations_created_date"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 MuiGrid-grid-md-4"
        >
          <span
            className="MuiTypography-root MuiTypography-caption"
          >
            relations_created_date
          </span>
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            -
          </h6>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root makeStyles-tabStyle-1158 MuiGrid-container"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-3"
        data-testid="last_modified_by"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 MuiGrid-grid-md-4"
        >
          <span
            className="MuiTypography-root MuiTypography-caption"
          >
            last_modified_by
          </span>
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            -
          </h6>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-9"
        data-testid="last_modified_date"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-4 MuiGrid-grid-md-4"
        >
          <span
            className="MuiTypography-root MuiTypography-caption"
          >
            last_modified_date
          </span>
          <h6
            className="MuiTypography-root MuiTypography-h6"
          >
            -
          </h6>
        </div>
      </div>
    </div>
  </SimpleForm>,
]
`;
